File: 7xBot.py
from dotenv import load_dotenv
import os
import asyncio
import json
import discord
from discord.ext import commands, tasks
import random
from datetime import datetime, timedelta
from collections import Counter
import re
import pytz

load_dotenv()

TOKEN = os.getenv("DISCORD_TOKEN")
if TOKEN is None:
    raise ValueError("DISCORD_TOKEN environment variable is not set.")

CHANNEL_ID = int(os.getenv("WYD_CHANNEL_ID") or 0)
if CHANNEL_ID == 0:
    raise ValueError("WYD_CHANNEL_ID environment variable is not set or invalid.")

SUMMARY_CHANNEL_ID = 1433550440649199879
SUMMARY_TRIGGER_COUNT = 25

TIMEZONE = pytz.timezone('America/New_York')

intents = discord.Intents.default()
intents.message_content = True
intents.members = True

bot = commands.Bot(command_prefix='!', intents=intents)

photo_session_active = False
submitted_users = set()
user_streaks = {}
STREAKS_FILE = "photo_streaks.json"
summary_message_count = 0
summary_lock = asyncio.Lock()

STOPWORDS = {
    "the", "and", "for", "that", "this", "with", "you", "your", "are", "but", "not", "have",
    "was", "were", "they", "their", "from", "just", "about", "what", "when", "where", "how",
    "why", "can", "cant", "could", "would", "should", "will", "im", "its", "dont", "did", "didnt",
    "got", "get", "like", "lol", "lmao", "bro", "nah", "yes", "no", "too", "very", "then",
    "there", "here", "into", "out", "our", "yall", "rn", "wyd", "send", "photo", "photos", "a",
    "an", "to", "of", "in", "on", "at", "is", "it", "be", "as", "or", "if", "we", "i", "u",
    "me", "my", "so", "up", "do", "does", "did", "had", "has"
}


def get_random_ping_time():
    now = datetime.now(TIMEZONE)
    target = now.replace(
        hour=random.randint(18, 22),
        minute=random.randint(0, 59),
        second=0,
        microsecond=0
    )
    # If hour is 22, cap minutes to keep it before 23:00
    if target.hour == 22:
        target = target.replace(minute=random.randint(0, 59))
    # If the time has already passed today, schedule for tomorrow
    if target <= now:
        target += timedelta(days=1)
    return target


def load_streaks():
    if not os.path.exists(STREAKS_FILE):
        return {}

    try:
        with open(STREAKS_FILE, "r", encoding="utf-8") as f:
            raw = json.load(f)
    except (json.JSONDecodeError, OSError):
        return {}

    if not isinstance(raw, dict):
        return {}

    parsed = {}
    for key, value in raw.items():
        try:
            uid = int(key)
            streak_value = int(value)
        except (TypeError, ValueError):
            continue

        parsed[uid] = max(0, streak_value)

    return parsed


def save_streaks(streaks):
    serializable = {str(uid): int(value) for uid, value in streaks.items()}
    with open(STREAKS_FILE, "w", encoding="utf-8") as f:
        json.dump(serializable, f, indent=2, sort_keys=True)


def build_activity_summary(messages):
    authors = Counter(msg.author.display_name for msg in messages)

    words = []
    for msg in messages:
        cleaned = re.findall(r"[A-Za-z']{3,}", msg.content.lower())
        words.extend([word for word in cleaned if word not in STOPWORDS])

    top_people = authors.most_common(3)
    top_words = Counter(words).most_common(5)
    media_count = sum(len(msg.attachments) for msg in messages)

    people_line = ", ".join([f"{name} ({count})" for name, count in top_people]) or "No clear leaders"
    words_line = ", ".join([word for word, _ in top_words]) or "No strong keywords"

    return (
        "📝 **Chat Summary (last 20 messages)**\n"
        f"- Active people: {people_line}\n"
        f"- Main topics/keywords: {words_line}\n"
        f"- Media shared: {media_count} file(s)"
    )


async def maybe_post_channel_summary(message):
    global summary_message_count

    if message.author.bot or message.channel.id != SUMMARY_CHANNEL_ID:
        return

    summary_message_count += 1
    if summary_message_count < SUMMARY_TRIGGER_COUNT:
        return

    async with summary_lock:
        if summary_message_count < SUMMARY_TRIGGER_COUNT:
            return

        recent_messages = []
        async for msg in message.channel.history(limit=75):
            if msg.author.bot:
                continue
            recent_messages.append(msg)
            if len(recent_messages) == SUMMARY_TRIGGER_COUNT:
                break

        if not recent_messages:
            summary_message_count = 0
            return

        summary_text = build_activity_summary(list(reversed(recent_messages)))
        await message.channel.send(summary_text)
        summary_message_count = 0


async def send_photo_prompt():
    global photo_session_active, submitted_users, user_streaks

    channel = bot.get_channel(CHANNEL_ID)
    if channel is None:
        print("Channel not found.")
        return
    if not isinstance(channel, discord.TextChannel):
        print("Channel is not a text channel.")
        return
    guild = channel.guild

    # Close the previous cycle before opening a new one.
    if photo_session_active:
        current_member_ids = {member.id for member in guild.members if not member.bot}

        # Keep streak records only for current non-bot members.
        user_streaks = {uid: user_streaks.get(uid, 0) for uid in current_member_ids}

        for uid in current_member_ids:
            if uid in submitted_users:
                user_streaks[uid] = user_streaks.get(uid, 0) + 1
            else:
                user_streaks[uid] = 0

        save_streaks(user_streaks)

    photo_session_active = True
    submitted_users = set()

    await channel.send(
        f"@everyone 📸 WYD RN SEND A PHOTOOOOO"
    )


@bot.event
async def on_ready():
    global user_streaks

    user_streaks = load_streaks()
    print(f"Loaded {len(user_streaks)} streak records.")
    print(f'Logged in as {bot.user}')
    scheduler.start()


@bot.event
async def on_message(message):
    global submitted_users

    if message.author.bot:
        return

    if photo_session_active and message.channel.id == CHANNEL_ID and message.attachments:
        valid_exts = {'jpg', 'jpeg', 'png', 'gif', 'webp', 'heic', 'mp4', 'mov'}
        has_media = any(
            attachment.filename.lower().rsplit('.', 1)[-1] in valid_exts
            for attachment in message.attachments
            if '.' in attachment.filename
        )

        if has_media and message.author.id not in submitted_users:
            submitted_users.add(message.author.id)
            preview_streak = user_streaks.get(message.author.id, 0) + 1
            await message.add_reaction('✅')
            await message.channel.send(
                f"Nice <@{message.author.id}>. If you keep this up, your streak will be **{preview_streak}** after the next prompt."
            )

    await maybe_post_channel_summary(message)

    await bot.process_commands(message)


@tasks.loop(minutes=1)
async def scheduler():
    global next_ping_time

    now = datetime.now(TIMEZONE)

    if now >= next_ping_time:
        await send_photo_prompt()
        next_ping_time = get_random_ping_time()
        print(f"Next ping scheduled for: {next_ping_time}")


@scheduler.before_loop
async def before_scheduler():
    global next_ping_time
    await bot.wait_until_ready()
    next_ping_time = get_random_ping_time()
    print(f"First ping scheduled for: {next_ping_time}")


# Admin command to manually trigger the photo prompt
@bot.command()
@commands.has_permissions(administrator=True)
async def forceprompt(ctx):
    await send_photo_prompt()


# Admin command to check who hasn't responded
@bot.command()
@commands.has_permissions(administrator=True)
async def pending(ctx):
    if not photo_session_active:
        await ctx.send("No active photo cycle right now.")
        return

    channel = bot.get_channel(CHANNEL_ID)
    if channel is None or not isinstance(channel, discord.TextChannel):
        await ctx.send("Photo channel is unavailable.")
        return

    member_ids = {member.id for member in channel.guild.members if not member.bot}
    missing_ids = sorted(member_ids - submitted_users)

    if missing_ids:
        mentions = ' '.join([f'<@{uid}>' for uid in missing_ids])
        await ctx.send(f"Still waiting on: {mentions}")
    else:
        await ctx.send("Everyone has submitted for this cycle.")


@bot.command()
async def streak(ctx, member: discord.Member = None):
    target = member or ctx.author
    current = user_streaks.get(target.id, 0)

    if photo_session_active and target.id in submitted_users:
        current += 1

    await ctx.send(f"🔥 {target.display_name}'s current streak: **{current}**")


bot.run(TOKEN)

File: 7xBot.js
require('dotenv').config();

const {
    Client,
    GatewayIntentBits,
    PermissionFlagsBits,
    ChannelType,
} = require('discord.js');
const fs = require('fs');

const TOKEN = process.env.DISCORD_TOKEN;
if (!TOKEN) throw new Error('DISCORD_TOKEN environment variable is not set.');

const CHANNEL_ID = process.env.WYD_CHANNEL_ID;
if (!CHANNEL_ID) throw new Error('WYD_CHANNEL_ID environment variable is not set or invalid.');
const GROQ_API_KEY = process.env.GROQ_API_KEY;
const GROQ_SUMMARY_MODEL = process.env.GROQ_SUMMARY_MODEL || 'llama-3.1-8b-instant';

const STREAKS_FILE = 'photo_streaks.json';
const TIMEZONE = 'America/New_York';

const client = new Client({
    intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
        GatewayIntentBits.GuildMembers,
    ],
});

let photoSessionActive = false;
let submittedUsers = new Set();
let userStreaks = {};
let nextPingTime = null;
let isSendingPrompt = false;

// Returns a Date whose local fields reflect the current time in the Eastern timezone.
function getNowET() {
    return new Date(new Date().toLocaleString('en-US', { timeZone: TIMEZONE }));
}

function getNextDailyPingTime(forceTomorrow = false) {
    const now = getNowET();
    const hour = Math.floor(Math.random() * 5) + 18; // 18–22
    const minute = Math.floor(Math.random() * 60);

    const target = new Date(now);
    target.setHours(hour, minute, 0, 0);

    // On startup, allow today if still upcoming. After a sent prompt, always use tomorrow.
    if (forceTomorrow || target <= now) {
        target.setDate(target.getDate() + 1);
    }

    return target;
}

function formatEtDate(date) {
    return date.toLocaleString('en-US', {
        timeZone: TIMEZONE,
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit',
        hour12: true,
    });
}

function loadStreaks() {
    if (!fs.existsSync(STREAKS_FILE)) return {};
    try {
        const raw = JSON.parse(fs.readFileSync(STREAKS_FILE, 'utf-8'));
        if (typeof raw !== 'object' || Array.isArray(raw)) return {};
        const parsed = {};
        for (const [key, value] of Object.entries(raw)) {
            const streakValue = parseInt(value, 10);
            if (!isNaN(streakValue)) {
                parsed[key] = Math.max(0, streakValue);
            }
        }
        return parsed;
    } catch {
        return {};
    }
}

function saveStreaks(streaks) {
    const serializable = {};
    for (const [uid, value] of Object.entries(streaks)) {
        serializable[uid] = parseInt(value, 10);
    }
    fs.writeFileSync(STREAKS_FILE, JSON.stringify(serializable, null, 2));
}

function buildSummaryInput(messages) {
    return messages
        .map(msg => {
            const name = msg.member?.displayName ?? msg.author.username;
            const content = (msg.content ?? '').trim();
            return `${name}: ${content}`;
        })
        .filter(line => line.trim().length > 3)
        .join('\n')
        .slice(0, 6000);
}

async function summarizeWithGroq(messages) {
    if (!GROQ_API_KEY) {
        return 'Groq summarizer is not configured. Set GROQ_API_KEY in your .env file.';
    }

    if (typeof fetch !== 'function') {
        return 'Groq summarizer requires Node.js 18+ (global fetch is unavailable).';
    }

    const chatText = buildSummaryInput(messages);
    if (!chatText) {
        return 'No meaningful text was found in the recent messages to summarize.';
    }

    try {
        const response = await fetch('https://api.groq.com/openai/v1/chat/completions', {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${GROQ_API_KEY}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                model: GROQ_SUMMARY_MODEL,
                temperature: 0.2,
                max_tokens: 120,
                messages: [
                    {
                        role: 'system',
                        content: 'You summarize group chats. Return exactly 1-2 concise sentences and keep slang/context intact.'
                    },
                    {
                        role: 'user',
                        content: `Summarize this Discord chat:\n\n${chatText}`,
                    },
                ],
            }),
        });

        const data = await response.json().catch(() => ({}));
        if (!response.ok) {
            const apiError = data?.error?.message || `HTTP ${response.status}`;
            return `Groq summarization failed: ${apiError}`;
        }

        const summary = data?.choices?.[0]?.message?.content?.trim();
        if (!summary) return 'Groq summarization returned an empty response.';

        return summary;
    } catch (err) {
        return `Groq summarization failed: ${err.message}`;
    }
}

async function buildActivitySummary(messages, scopeLabel) {
    const authorCounts = {};
    for (const msg of messages) {
        const name = msg.member?.displayName ?? msg.author.username;
        authorCounts[name] = (authorCounts[name] ?? 0) + 1;
    }

    const topPeople = Object.entries(authorCounts)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 3)
        .map(([name, count]) => `${name} (${count})`);

    const peopleLine = topPeople.join(', ') || 'No clear leaders';

    const summaryLine = await summarizeWithGroq(messages);

    return (
        `📝 **Chat Summary (${scopeLabel})**\n` +
        `- Active people: ${peopleLine}\n` +
        `- Summary: ${summaryLine}`
    );
}

async function fetchRecentUserMessagesByCount(channel, count) {
    const target = Math.max(1, count);
    const out = [];
    let before;

    while (out.length < target) {
        const remaining = target - out.length;
        const batchSize = Math.min(100, Math.max(remaining * 2, 25));
        const fetched = await channel.messages.fetch({ limit: batchSize, before });
        if (!fetched.size) break;

        for (const msg of fetched.values()) {
            if (msg.author.bot) continue;
            out.push(msg);
            if (out.length === target) break;
        }

        before = fetched.last()?.id;
        if (!before) break;
    }

    return out.reverse();
}

async function fetchRecentUserMessagesSince(channel, sinceDate) {
    const out = [];
    let before;

    while (true) {
        const fetched = await channel.messages.fetch({ limit: 100, before });
        if (!fetched.size) break;

        let reachedOlderThanWindow = false;
        for (const msg of fetched.values()) {
            if (msg.createdAt < sinceDate) {
                reachedOlderThanWindow = true;
                continue;
            }
            if (!msg.author.bot) out.push(msg);
        }

        if (reachedOlderThanWindow) break;

        before = fetched.last()?.id;
        if (!before) break;
    }

    return out.reverse();
}

function parseSummaryWindow(args) {
    if (!args.length) {
        return {
            ok: true,
            mode: 'count',
            count: 30,
            scopeLabel: 'last 30 messages',
        };
    }

    const [firstRaw, secondRaw] = args;
    const first = firstRaw.toLowerCase();
    const second = (secondRaw ?? '').toLowerCase();

    if (['30', '60', '120'].includes(first) && args.length === 1) {
        const count = parseInt(first, 10);
        return {
            ok: true,
            mode: 'count',
            count,
            scopeLabel: `last ${count} messages`,
        };
    }

    let amount;
    let unit;

    const compact = first.match(/^(\d+)\s*([a-z]+)$/i);
    if (compact && args.length === 1) {
        amount = parseInt(compact[1], 10);
        unit = compact[2].toLowerCase();
    } else if (/^\d+$/.test(first) && args.length >= 2) {
        amount = parseInt(first, 10);
        unit = second;
    }

    if (!amount || amount <= 0 || !unit) {
        return {
            ok: false,
            error: 'Usage: `!summary 30|60|120` or `!summary <number> <m|min|h|hr|d|day>` (example: `!summary 25 m`, `!summary 2 hr`).',
        };
    }

    let multiplierMs;
    if (['m', 'min', 'mins', 'minute', 'minutes'].includes(unit)) {
        multiplierMs = 60_000;
    } else if (['h', 'hr', 'hrs', 'hour', 'hours'].includes(unit)) {
        multiplierMs = 3_600_000;
    } else if (['d', 'day', 'days'].includes(unit)) {
        multiplierMs = 86_400_000;
    } else {
        return {
            ok: false,
            error: 'Unknown time unit. Use minutes (`m`), hours (`hr`), or days (`d`).',
        };
    }

    const since = new Date(Date.now() - amount * multiplierMs);
    const unitLabel = unit;
    return {
        ok: true,
        mode: 'time',
        since,
        scopeLabel: `since ${amount} ${unitLabel} ago`,
    };
}

async function sendPhotoPrompt() {
    const channel = await client.channels.fetch(CHANNEL_ID).catch(() => null);
    if (!channel || channel.type !== ChannelType.GuildText) {
        console.log('Channel not found or not a text channel.');
        return false;
    }
    const guild = channel.guild;

    if (photoSessionActive) {
        await guild.members.fetch();
        const currentMemberIds = guild.members.cache
            .filter(m => !m.user.bot)
            .map(m => m.id);

        const newStreaks = {};
        for (const uid of currentMemberIds) {
            // Streaks are awarded immediately on upload; only reset misses at cycle boundary.
            if (!submittedUsers.has(uid)) {
                newStreaks[uid] = 0;
            } else {
                newStreaks[uid] = userStreaks[uid] ?? 0;
            }
        }
        userStreaks = newStreaks;
        saveStreaks(userStreaks);
    }

    photoSessionActive = true;
    submittedUsers = new Set();

    await channel.send('@everyone 📸 WYD RN SEND A PHOTOOOOO');
    return true;
}

client.once('ready', () => {
    userStreaks = loadStreaks();
    console.log(`Loaded ${Object.keys(userStreaks).length} streak records.`);
    console.log(`Logged in as ${client.user.tag}`);

    nextPingTime = getNextDailyPingTime();
    console.log(`First ping scheduled for (ET): ${formatEtDate(nextPingTime)}`);

    setInterval(async () => {
        const now = getNowET();
        if (isSendingPrompt) return;
        if (now >= nextPingTime) {
            isSendingPrompt = true;
            try {
                const sent = await sendPhotoPrompt();
                if (sent) {
                    nextPingTime = getNextDailyPingTime(true);
                    console.log(`Next ping scheduled for (ET): ${formatEtDate(nextPingTime)}`);
                } else {
                    // Keep cadence daily-only. If a send fails, do not create extra cycles.
                    nextPingTime = getNextDailyPingTime(true);
                    console.log(`Prompt send failed; next daily ping remains (ET): ${formatEtDate(nextPingTime)}`);
                }
            } catch (err) {
                console.error('Failed to send photo prompt:', err);
                nextPingTime = getNextDailyPingTime(true);
                console.log(`Send errored; next daily ping remains (ET): ${formatEtDate(nextPingTime)}`);
            } finally {
                isSendingPrompt = false;
            }
        }
    }, 60_000);
});

client.on('messageCreate', async (message) => {
    if (message.author.bot) return;

    if (photoSessionActive && message.channel.id === CHANNEL_ID && message.attachments.size > 0) {
        const validExts = new Set(['jpg', 'jpeg', 'png', 'gif', 'webp', 'heic', 'mp4', 'mov']);
        const hasMedia = message.attachments.some(att => {
            const fileName = att.filename?.toLowerCase() ?? '';
            const ext = fileName.includes('.') ? fileName.split('.').pop() : '';
            const contentType = (att.contentType ?? '').toLowerCase();
            return validExts.has(ext) || contentType.startsWith('image/') || contentType.startsWith('video/');
        });

        if (hasMedia && !submittedUsers.has(message.author.id)) {
            submittedUsers.add(message.author.id);
            userStreaks[message.author.id] = (userStreaks[message.author.id] ?? 0) + 1;
            saveStreaks(userStreaks);
            await message.react('✅');
            await message.channel.send(
                `Nice <@${message.author.id}>. Your streak is now **${userStreaks[message.author.id]}**.`
            );
        }
    }

    // Command handling
    const PREFIX = '!';
    if (!message.content.startsWith(PREFIX)) return;

    const args = message.content.slice(PREFIX.length).trim().split(/\s+/);
    const command = args.shift().toLowerCase();

    if (command === 'forceprompt') {
        if (!message.member.permissions.has(PermissionFlagsBits.Administrator)) return;
        await sendPhotoPrompt();

    } else if (command === 'nextping') {
        if (!message.member.permissions.has(PermissionFlagsBits.Administrator)) return;
        if (!nextPingTime) {
            await message.channel.send('No ping is currently scheduled.');
            return;
        }
        await message.channel.send(`Next prompt is scheduled for **${formatEtDate(nextPingTime)} ET**.`);

    } else if (command === 'pending') {
        if (!message.member.permissions.has(PermissionFlagsBits.Administrator)) return;

        if (!photoSessionActive) {
            await message.channel.send('No active photo cycle right now.');
            return;
        }

        const channel = client.channels.cache.get(CHANNEL_ID);
        if (!channel || channel.type !== ChannelType.GuildText) {
            await message.channel.send('Photo channel is unavailable.');
            return;
        }

        await channel.guild.members.fetch();
        const memberIds = channel.guild.members.cache
            .filter(m => !m.user.bot)
            .map(m => m.id);

        const missingIds = memberIds.filter(id => !submittedUsers.has(id)).sort();

        if (missingIds.length) {
            const mentions = missingIds.map(id => `<@${id}>`).join(' ');
            await message.channel.send(`Still waiting on: ${mentions}`);
        } else {
            await message.channel.send('Everyone has submitted for this cycle.');
        }

    } else if (command === 'streak') {
        let target = message.member;

        if (args.length > 0) {
            const id = args[0].replace(/[<@!>]/g, '');
            const found = message.guild.members.cache.get(id);
            if (found) target = found;
        }

        const current = userStreaks[target.id] ?? 0;

        await message.channel.send(`🔥 ${target.displayName}'s current streak: **${current}**`);

    } else if (command === 'summary') {
        const parsed = parseSummaryWindow(args);
        if (!parsed.ok) {
            await message.channel.send(parsed.error);
            return;
        }

        let messagesToSummarize = [];
        if (parsed.mode === 'count') {
            messagesToSummarize = await fetchRecentUserMessagesByCount(message.channel, parsed.count);
        } else {
            messagesToSummarize = await fetchRecentUserMessagesSince(message.channel, parsed.since);
        }

        if (!messagesToSummarize.length) {
            await message.channel.send('none');
            return;
        }

        const summaryText = await buildActivitySummary(messagesToSummarize, parsed.scopeLabel);
        await message.channel.send(summaryText);
    }
});

client.login(TOKEN);
